Mathlib/LinearAlgebra/Goursat.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.LinearAlgebra.Prod
+import Mathlib.LinearAlgebra.Quotient.Basic
+import Mathlib.GroupTheory.Goursat
+
+/-!
+# Goursat's lemma for submodules
+
+Let `G, H` be modules over a ring `R`. If `I` is a submodule of `G × H` which projects fully on
+to both factors, then there exist submodules `G' ≤ G` and `H' ≤ H` such that `G' × H' ≤ I` and the
+image of `I` in `(G ⧸ G') × (H ⧸ H')` is the graph of an isomorphism `G ⧸ G' ≃ₗ[R] H ⧸ H'`.
+
+`G'` and `H'` can be explicitly constructed as `Submodule.goursatFst I` and `Submodule.goursatSnd I`
+respectively.
+-/
+
+open Function Set LinearMap
+
+namespace Submodule
+variable {R G H : Type*} [Ring R] [AddCommGroup G] [Module R G] [AddCommGroup H] [Module R H]
+  {I : Submodule R (G × H)}
```

Comment:
Can `G` and `H` be renamed to `M` and `N` everywhere since they are modules?

Reply:
Done.